Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/model/ConversionMap.kt

```diff
@@ -0,0 +1,451 @@
+    }
+
+    fun getConversions(fromType: DataType): MutableList<Conversion> {
+        return map.computeIfAbsent(fromType) { ArrayList() }
+    }
+
+    /*
+    Returns conversions for the given type, or any supertype, recursively
+     */
+    private fun getAllConversions(fromType: DataType?): List<Conversion> {
+        val conversions: MutableList<Conversion> = ArrayList()
+        var currentType = fromType
+        while (currentType != null) {
+            conversions.addAll(getConversions(currentType))
+            currentType = currentType.baseType
+        }
+        return conversions
+    }
+
+    private fun findCompatibleConversion(fromType: DataType, toType: DataType): Conversion? {
+        if (fromType.isCompatibleWith(toType)) {
+            return Conversion(fromType, toType)
+        }
+
+        return null
+    }
+
+    private fun findChoiceConversion(
+        fromType: ChoiceType,
+        toType: DataType,
+        allowPromotionAndDemotion: Boolean,
+        operatorMap: OperatorMap
+    ): Conversion? {
+        var result: Conversion? = null
+        for (choice in fromType.types) {
+            val choiceConversion =
+                findConversion(choice, toType, true, allowPromotionAndDemotion, operatorMap)
+            if (choiceConversion != null) {
+                if (result == null) {
+                    result = Conversion(fromType, toType, choiceConversion)
+                } else {
+                    result.addAlternativeConversion(choiceConversion)
+                }
+            }
+        }
+
+        return result
+    }
+
+    private fun findTargetChoiceConversion(
+        fromType: DataType,
+        toType: ChoiceType,
+        allowPromotionAndDemotion: Boolean,
+        operatorMap: OperatorMap
+    ): Conversion? {
+        for (choice in toType.types) {
+            findConversion(fromType, choice, true, allowPromotionAndDemotion, operatorMap)?.let {
+                return Conversion(fromType, toType, it)
+            }
+        }
+
+        return null
+    }
+
+    private fun findListConversion(
+        fromType: ListType,
+        toType: ListType,
+        operatorMap: OperatorMap
+    ): Conversion? {
+        return findConversion(
+                fromType.elementType,
+                toType.elementType,
+                isImplicit = true,
+                allowPromotionAndDemotion = false,
+                operatorMap = operatorMap
+            )
+            ?.let { Conversion(fromType, toType, it) }
+    }
+
+    private fun findIntervalConversion(
+        fromType: IntervalType,
+        toType: IntervalType,
+        operatorMap: OperatorMap
+    ): Conversion? {
+        return findConversion(
+                fromType.pointType,
+                toType.pointType,
+                isImplicit = true,
+                allowPromotionAndDemotion = false,
+                operatorMap = operatorMap
+            )
+            ?.let { Conversion(fromType, toType, it) }
```

Comment:
Question, is this the equivalent of Optional.ofNullable(findConversion(..)).map(it -> new Conversion(...)

Reply:
Yes. `.let` only runs if the result is not null.